lib/text/ssa_text_parser.js

```diff
@@ -0,0 +1,310 @@
+/*! @license
+ * Shaka Player
+ * Copyright 2016 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+goog.provide('shaka.text.SsaTextParser');
+
+goog.require('goog.asserts');
+goog.require('shaka.log');
+goog.require('shaka.text.Cue');
+goog.require('shaka.text.TextEngine');
+goog.require('shaka.util.StringUtils');
+
+
+/**
+ * Documentation: http://moodub.free.fr/video/ass-specs.doc
+ * https://en.wikipedia.org/wiki/SubStation_Alpha
+ * @implements {shaka.extern.TextParser}
+ * @export
+ */
+shaka.text.SsaTextParser = class {
+  /**
+   * @override
+   * @export
+   */
+  parseInit(data) {
+    goog.asserts.assert(false, 'SSA does not have init segments');
+  }
+
+  /**
+   * @override
+   * @export
+   */
+  parseMedia(data, time) {
+    return shaka.text.SsaTextParser.getCues_(data);
+  }
+
+  /**
+   * @param {BufferSource} data
+   * @return {!Array.<!shaka.extern.Cue>}
+   * @private
+   */
+  static getCues_(data) {
+    const StringUtils = shaka.util.StringUtils;
+    const SsaTextParser = shaka.text.SsaTextParser;
+
+    // Get the input as a string.
+    const str = StringUtils.fromUTF8(data);
+
+
+    const styles = [];
+
+    /** @type {!Array.<!shaka.extern.Cue>} */
+    const cues = [];
+    const parts = str.split(/\r?\n\s*\r?\n/);
+    for (const part of parts) {
```

Comment:
I think it would be easier to understand if you split the raw parsing and the name handling into separate parts.  For example, have a parse method that produces a nested object map:

```json
{
  "Script Info": {
    "Title": "Foo",
    "Original Script": ""
  },
  "V4+ Styles": {
    "Format": "..."
  }
}
```

Reply:
I'm going to think of a way to make it easier, and hope to come back tomorrow with a different implementation.

Reply:
Done!

Reply:
I was suggesting a more generic parser that didn't check for tag names.  For example, something like:

```js
const ret = {};
let section = null;
for (const line of lines) {
  const match = /\[([^\]]+)\]/.exec(line);
  if (match) {
    section = match[1];
    ret[section] = {};
  } else if (/\s*;/.test(line)) {
    // Skip comments.
  } else {
    const parts = line.split(':');
    ret[section][parts[0].trim()] = parts.slice(1).join(':').trim();
    // You'll need to special-case Dialogue to support multiple cues.
  }
}
```

I would also suggest putting this in another function to convert the data buffer to the object.  Then you can just use `data['V4 Styles']` or `data['Events']` later.

Reply:
Given that it is a popular format, what is the advantage of doing what you propose? At this time, what is supported are known tags and those that we add to sections. Adding unknown sections I don't think has any advantage ... Sorry if I didn't understand you.

Reply:
Your code "adds" unknown sections too by ignoring them in the if statements below.  I was suggesting we separate the text parsing from the handling of the sections.  Like you would split XML parsing from the reading of the elements.  By splitting them up, you can focus on reading the tag names without seeing all the regex and string parsing.

Reply:
Sorry, I still don't see it :(. Right now with the last changes the events and styles are already separated, and also the parsing of them is made independent. Adding a generic parse here does not add much since it is a known format.In your example above, script information is added to the object that is not really useful.
On the other hand, even if I do a generic parse, I still have to know the tags.
The execution time of my proposal and yours should be similar.